src/lang/base/Syntax.ml

```diff
@@ -499,6 +500,63 @@ module ScillaSyntax (SR : Rep) (ER : Rep) = struct
         let body' = subst_type_in_expr tvar tp body in
         (Fixpoint (f, t', body'), rep)
 
+(* Is expr dependent on any ident in blist.
+ * This is the same as checking if a free var
+ * in expr is present in blist. *)
+  let free_vars_dep_check erep blist =
+
+    (* is m in l. *)
+    let is_mem m l =
+      List.exists l ~f:(fun x -> get_id m = get_id x) in
+    (* is any m in ml, in l. *)
+    let any_is_mem ml l =
+      List.exists ml ~f:(fun i -> is_mem i l) in
+    (* get elements in "l" that are not in bound_vars. *)
+    let get_free l bound_vars =
+      List.filter l ~f:(fun i -> not (is_mem i bound_vars)) in
+    let rec get_pattern_bounds p =
+      match p with
+      | Wildcard -> []
+      | Binder i -> [i]
+      | Constructor (_, plist) ->
+        List.fold plist ~init:[] ~f:(fun acc p' ->
+          let l = get_pattern_bounds p' in
+          acc @ l)
```

Comment:
Can we add an accumulator to get_pattern_bounds, or even better define it to use a local function that uses an accumulator? The expression `acc @ l` is linear in the size of `acc`, so `get_pattern_bounds` ends up being quadratic in the size of the pattern rather than linear.